Fixes #21120 - orchestration tasks are added only once

lzap opened the pull request:
Previously, our orchestration task engine was accepting multiple instances of same tasks. Foreman is then orchestrating various actions multiple times and this misbehavior is with us for years. All our backend services were idempotent, until we added PXELoader flag which caused DHCP to error out with conflicts, because filename option treated separately in ISC DHCP.

Without this patch, DHCP orchestration of host edit had four steps:

* delete DHCP entry
* delete DHCP entry
* add DHCP entry
* add DHCP entry

This patch will not add orchestration tasks with the same name twice. If you grep in our codebase for `queue.create` you can see that we don't have any task with same name, so `name` is good field to check for. I considered checking `action` but it contain active record reference which changes over time.

Now, `name` is a free form string and it is also being translated, so it is not ideal. There might be cases when multiple tasks must be scheduled. In that case, new field was introduced called `id` and if that's supplied (symbol or string), it is considered as the identifier for searching for duplicite entries.

With the patch, it's now correct:

* delete DHCP entry
* add DHCP entry

If needed, I can also add `id` field to all core tasks and add a deprecation warning for plugins which does not provide it so plugin authors have time to fix their code until we enforce presence of id.

lzap committed:
Fixes #21120 - orchestration tasks are added only once

lzap commented on app/services/orchestration/queue.rb:
```diff
@@ -17,16 +17,30 @@ def initialize(name = "Unnamed")
+    end
+
+    def find_by_id(id)
+      items.find {|task| task.id == id}
```
I knew it! Damn.

ares commented on test/models/orchestration/dhcp_test.rb:
```diff
@@ -239,9 +239,9 @@ def teardown
-    # 1st is creation from factory, 2nd is triggered by h.valid?
-    assert_equal 2, h.queue.items.count {|x| x.action == [ h.primary_interface, :set_dhcp ] }
-    # and also one deletion (of original creation)
+    # TODO replace this with h.queue.find_by_action - I want these changes to be more explicit for the initial review,
```
I guess you can now replace it :-) makes sense to me

lzap replied:
Sure, anything else before I rebase?

lzap replied:
@ares can you confirm that the changes in this test (2 -> 1) are correct? I want just double check this, I don't see any reason to have these, I think it was simply verified because of the design. I guess "makes sense" is that, want to be sure. Possibly breaking change it is.

lzap replied:
I will implement "IDs" (see initial description) which will make this much easier. Matching via actions is not possible because NIC/Host are not kept in memory for whole time.

ares replied:
I believe this is correct, I believe I bumped those numbers just because `valid?` added new item to queue

ares commented on app/services/orchestration/queue.rb:
```diff
@@ -38,6 +52,10 @@ def clear
       true
     end
 
+    def count
```
size alias would also be nice

ares commented on app/services/orchestration/task.rb:
```diff
@@ -24,11 +26,12 @@ def status?(symbol)
 
     def to_s
-      "#{name}\t #{priority}\t #{status}\t #{action}"
+      tid = id.nil? ? '' : " (#{id})"
```
woud it be better to rather have `(no id)` or something in case there's no id set? hopefully that would indicate, that task author should consider setting id

lzap replied:
Yeah, I actually plan to do DEPRECATED warning in case of missing id, that should be explicit enough I think.

lzap replied:
New version has `id` required now, deprecation warning raised when it's missing. I want to keep both ID and name here just to be clear about this.

timogoebel replied:
Can't we generate the id automatically by default, based on the task name, action class name, action method, action object id, priority, ... maybe using a hash?

This seems to be a lot of change effort in plugins just to detect duplicates.
Why doesn't something as simple as `[task1, task2, task3].uniq` work?
Or clearing the queue(s) before new tasks are added to make sure they are only added once?

ares replied:
I'm not sure how many plugins extend the orchestration but if there are some, we should update them too so they pass id, otherwise it won't happen until we hard enforce it since all plugins would rely on defaulting to task name or action method. Since errors caused by duplicated tasks in queue are hard to debug, I'd prefer 500 unless that means updating all foreman plugins. If we talk about ~5, I think it's worth it

ares commented on package.json:
```diff
@@ -7,7 +7,7 @@
     "@storybook/addon-actions": "^3.2.12",
     "@storybook/react": "^3.2.12",
-    "@storybook/storybook-deployer": "^2.0.0",
+    "@storybook/storybook-deployer": "^2.2.0",
```
changes in this file seems unrelated?

lzap replied:
Will remove, thanks.

ares replied:
still here

ares commented on app/services/orchestration/task.rb:
```diff
@@ -41,5 +44,10 @@ def update_ts
+
+    def ==(other)
+      return self.id == other.id if self.id && other.id
+      self.name == other.name
```
I didn't dig into details yet but my understanding is that we only compare id or name if id is missing, originally we discussed we should also look at all attributes/parameters, is it no longer necessary?

lzap replied:
I can't compare parameters, there is usually whole object passed in (host, nic) that is modified. It's either ID (recommended) or name (deprecated, will be removed in 1.18).

lzap replied:
In the next Foreman version, ID will be required. The fallback on name won't be used anymore. We can't easily compare data as it contains ActiveRecord objects.

ares reviewed (changes requested):
I left several comments from initial review

should this PR also extend existing orchestration and set id for tasks?

lzap commented:
Let me know what you think @ares and @ohadlevy - the new version:

* Adds new key called `id` which is now required and used as ident to recognize dupes. When not present, deprecation warning is shown and `name` is used as a fallback, which is fine for most cases (but for NICs this can be problem - I already added `id` there).
* Tasks are now preserved in stable order - what comes first with same priority comes out. On my Ruby the sorting algo seems to be stable, but I cannot guarantee other versions, thus I added flag `created` and made sure tasks are preserved in this order after sort.
* Added some new helper methods to orchestration queue which makes tests cleaner.
* Added new unit test for orchestration queue.
* Rewrote dhcp orchestration test with new approach which uses `id` instead of clunky testing of actions.

The last step (test rerwrite) revealed also that we have wrong order for new hosts which are also being edited. The correct order would be: create new entry, edit the entry (delete, create). But we have it the other way around, because create is priority 10 and edit is 5/6. I think since this patch is touching lots of things and adds the stable sort, I propose to change the create DHCP priority from 10 to 3.

To test the misbehavior I am trying to fix just edit Orchestration::Queue#create in a way that multiple entries are allowed, then run tests and see:

```
Failure:
DhcpOrchestrationTest#test_0017_when an existing host triggers a 'rebuild', its dhcp records should be updated if no dhcp records are found [test/models/orchestration/dhcp_test.rb:287]:
--- expected
+++ actual
@@ -1 +1 @@
-["dhcp_remove_aa:bb:cc:dd:ee:f1", "dhcp_create_aa:bb:cc:dd:ee:f1"]
+["dhcp_remove_aa:bb:cc:dd:ee:f1", "dhcp_create_aa:bb:cc:dd:ee:f1", "dhcp_create_aa:bb:cc:dd:ee:f1"]

Failure:
DhcpOrchestrationTest#test_0014_when an existing host changes its ip address, its dhcp records should be updated [test/models/orchestration/dhcp_test.rb:245]:
--- expected
+++ actual
@@ -1 +1 @@
-["dhcp_remove_aa:bb:cc:dd:ee:ff", "dhcp_create_aa:bb:cc:dd:ee:ff"]
+["dhcp_remove_aa:bb:cc:dd:ee:ff", "dhcp_create_aa:bb:cc:dd:ee:ff", "dhcp_create_aa:bb:cc:dd:ee:ff"]

Failure:
DhcpOrchestrationTest#test_0011_new host with multiple validations should create a single dhcp reservation [test/models/orchestration/dhcp_test.rb:209]:
--- expected
+++ actual
@@ -1 +1 @@
-["dhcp_create_aa:bb:cc:dd:ee:ff"]
+["dhcp_create_aa:bb:cc:dd:ee:ff", "dhcp_create_aa:bb:cc:dd:ee:ff"]

... etc
```

Not all test cases will fail, obviously only those when we call `valid?` multiple times. I created an extra test when I explicitly call this twice to check for this case.

I also created alternative backport-friendly version of this approach that can go to 1.15/1.16 because this is too much of a change I think. See the referenced PR. When testing this, be aware that smart-proxy from develop currently ignores filename option and does not serialize it: http://projects.theforeman.org/issues/21975

lzap commented:
@ares let me know and I will rebase, after branching I'd love to have this merged so we have more testing.

ares commented on app/services/orchestration/task.rb:
```diff
@@ -39,7 +45,14 @@ def update_ts
+    def ==(other)
+      return false unless other.is_a?(Task)
+      return self.id == other.id if self.id && other.id
+      self.name == other.name
```
if `id` is now required, why we fallback to name?

ares replied:
now I get it after reading your last comment, can't we then use name in initializer as an id if id was not passed as argument? therefore we wouldn't need fallback here (and potentially other places)

lzap replied:
Absolutely good idea that's also faster!

ares commented:
left few comments, I guess you can rebase, also see @timogoebel's questions, but from my point of view, this is going right way

lzap commented:
Rebased, I know that @ares is preparing FOSDEM demo now, @timogoebel might be away too. I would love to get that merged soon as we branched off 1.17 and this is quite a change. Maybe @dLobatog, @tbrisker or @ShimShtein can review and test this? Thanks a bunch

lzap commented on test/models/orchestration/dhcp_test.rb:
```diff
@@ -326,7 +328,7 @@ def teardown
     assert h.valid?
     assert bmc.valid?
-    assert_equal 1, h.queue.items.count {|x| x.action == [ h.primary_interface, :set_dhcp ] }
+    assert_equal 2, h.queue.items.count {|x| x.action == [ h.primary_interface, :set_dhcp ] }
```
Rebased, jeeeeez.

lzap commented on test/models/orchestration/dhcp_test.rb:
```diff
@@ -326,9 +326,9 @@ def teardown
+    assert_equal 2, h.queue.items.count {|x| x.action == [ h.primary_interface, :set_dhcp ] }
     assert_equal 1, h.queue.items.count {|x| x.action.last == :del_dhcp }
-    assert_equal 1, bmc.queue.items.count {|x| x.action == [ bmc,     :set_dhcp ] }
+    assert_equal 1, bmc.queue.items.count {|x| x.action == [ bmc, :set_dhcp ] }
```
I fixed this already, local rubocop is okay with the fix. Vanish!

lzap commented on test/models/orchestration/dhcp_test.rb:
```diff
@@ -326,10 +326,10 @@ def teardown
+    assert_equal 2, h.queue.items.count {|x| x.action == [ h.primary_interface, :set_dhcp ]}
+    assert_equal 1, h.queue.items.count {|x| x.action.last == :del_dhcp}
+    assert_equal 1, bmc.queue.items.count {|x| x.action == [ bmc, :set_dhcp ]}
+    assert_equal 1, bmc.queue.items.count {|x| x.action == [ bmc.old, :del_dhcp ]}
```
Are u fing king me?

ares commented on app/services/orchestration/queue.rb:
```diff
@@ -17,22 +17,37 @@ def initialize(name = "Unnamed")
     end
 
+    def task_ids
+      all.map{|x| x.id}
```
is better to be written as `all.map(&:id)`

lzap replied:
Fixed, rebased.

ares commented on app/services/orchestration/task.rb:
```diff
@@ -1,12 +1,17 @@
       @action    = opts[:action]
+      @created   = opts[:created] || Time.now.to_f
+      raise("action must be present for task '#{name}'") if action.nil?
+      # this kills tests, need to update other tests before
```
what tests you had in mind? and why action is now required? I mean is it refactoring so we fail faster if action is missing?

lzap replied:
The comment "this kills" was meant for line 14, deprecated exceptions actually make tests to fail. Re: action - since all attributes are read only, it makes no sense to create such a task?

ares commented on app/services/orchestration/task.rb:
```diff
@@ -1,12 +1,17 @@
+      @created   = opts[:created] || Time.now.to_f
+      raise("action must be present for task '#{name}'") if action.nil?
+      # this kills tests, need to update other tests before
+      #Foreman::Deprecation.deprecation_warning('1.19', "Orchestration task '#{name}' is missing required :id field") if @id.nil?
```
should this still be commented?

lzap replied:
I removed those, my initial idea was to fail fast with deprecation warning to enforce having `id` but I think nobody will notice this in test output, it's better I think to have sane default (use name if id is not set) and go with that.

ares reviewed (changes requested):
I tested on nightly prod-like setup, orchestration still works great (also with several popular plugins). 3 nits in code and then good to be merged!

lzap commented:
```diff
diff --git a/app/services/orchestration/queue.rb b/app/services/orchestration/queue.rb
index 0e6575d48..c330b1614 100644
--- a/app/services/orchestration/queue.rb
+++ b/app/services/orchestration/queue.rb
@@ -45,7 +45,7 @@ module Orchestration
     end
 
     def task_ids
-      all.map{|x| x.id}
+      all.map(&:id)
     end
 
     def clear
diff --git a/app/services/orchestration/task.rb b/app/services/orchestration/task.rb
index 32f94cb0b..7fe4b442c 100644
--- a/app/services/orchestration/task.rb
+++ b/app/services/orchestration/task.rb
@@ -10,8 +10,6 @@ module Orchestration
       @action    = opts[:action]
       @created   = opts[:created] || Time.now.to_f
       raise("action must be present for task '#{name}'") if action.nil?
-      # this kills tests, need to update other tests before
-      #Foreman::Deprecation.deprecation_warning('1.19', "Orchestration task '#{name}' is missing required :id field") if @id.nil?
       update_ts
     end
```

ares commented:
All is good but jenkins seems to be failing on unrelated problem with phantomjs. I'd rather wait with merging until it's resolved.

ares commented:
[test foreman]

ares commented:
[test katello]

ares commented:
Thanks @lzap, merging!
